### PR TITLE
Group api&refactor

### PR DESCRIPTION
**Pull request by lolmaxz**
Feel free to leave comments.
@Huijiro

### CODE REVIEW

File: src/requests/AuthApi.ts
Comment (Huijiro) [inline src/requests/AuthApi.ts:58]: If these parameters are not optional you should specify them so in the Type in this case.

File: src/requests/AuthApi.ts
Comment (Huijiro) [inline src/requests/AuthApi.ts:77]: Optional: You can probably do these on a loop.

File: src/requests/BaseApi.ts
Comment (Huijiro) [inline src/requests/BaseApi.ts:31]: You don't need to pass queryOptions like that, if it is empty it will not affect the URL in any way if it is a URLSearchParams type.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:45]: Promise.reject and throw are the same thing, you can throw here instead if you want.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:62]: You can probably do this with a ENUM instead of verifying in code.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:79]: This should be a ENUM as well.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:141]: Optional: There's a way to rewrite this better, this works, but it's a challenge for you to try to make this differently.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:169]: Another ENUM.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:1]: In general in this file there's a lot of Promise.reject that should just be throws.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:328]: There's a way better way to do this, i would throw a error here if it's not in range and define default separetly.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:334]: This should throw.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:392]: Same thing here.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:507]: You can compose this differently, remember that anything that is passed and is undefined is just going to be empty when create an object with properties.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [inline src/requests/GroupsApi.ts:557]: Again same problem here, the code is the same 3 times already...

File: src/requests/GroupsApi.ts
Comment (Huijiro) [reply src/requests/GroupsApi.ts:1]: The problems show here repeat all over the file, so take a look at them. Overall way better than before.

File: src/requests/BaseApi.ts
Comment (Huijiro) [inline src/requests/BaseApi.ts:138]: This may become a issue in the future if you really think about it.

File: src/requests/UsersApi.ts
Comment (Huijiro) [inline src/requests/UsersApi.ts:105]: Type checking on typescript...

File: src/requests/UsersApi.ts
Comment (Huijiro) [inline src/requests/UsersApi.ts:156]: ???? Really?

Comment (Huijiro) [reply src/requests/UsersApi.ts:156]: Think really hard on what you wrote here.

File: src/requests/UsersApi.ts
Comment (Huijiro) [inline src/requests/UsersApi.ts:172]: If you solve the problem of the getVRCRankTag function you will see that this here is useless.

File: src/requests/UsersApi.ts
Comment (Huijiro) [inline src/requests/UsersApi.ts:191]: This is alright, i don't know much about how VRC works, but try not to be opnionated on the way you do it, maybe a Nuisance rank player is not a troll, just a child etc, you could have the opnion most of them are probably, but if the system has a tag for that you should just use what the system has.

Comment (Huijiro) [reply src/requests/UsersApi.ts:191]: Remember other people will use the Wrapper, they may have different opnions than you.

File: src/VRCWrapper.ts
Comment (Huijiro) [inline src/VRCWrapper.ts:192]: If you throw the error and the error has a log in their constructor you should not need this.

File: src/VRCWrapper.ts
Comment (Huijiro) [inline src/VRCWrapper.ts:198]: This can be handle by checking the error code of the request.

File: src/VRCWrapper.ts
Comment (Huijiro) [inline src/VRCWrapper.ts:64]: All these errors should be in the constructors of the errors so they log these when they are Throw.

File: src/WebSocketHandler.ts
Comment (Huijiro) [inline src/WebSocketHandler.ts:1]: I will wait until you're done with this to review it.

File: src/errors.ts
Comment (Huijiro) [inline src/errors.ts:1]: Going the right way here by replacing their messages, but the set prototype of is still an issue.

**Review by Huijiro — CHANGES_REQUESTED** [review]
There's a few issues to be solved, mostly more complex stuff you need to re-think and some optimizing, but overall good.

File: src/requests/AuthApi.ts
Comment (lolmaxz) [reply src/requests/AuthApi.ts:58]: they are all optional, but at least one is required, if none is given, we send an error. We can give multiple at the same time, but minimum one parameter must be given.

File: src/requests/GroupsApi.ts
Comment (lolmaxz) [reply src/requests/GroupsApi.ts:62]: I'm not sure I understand because GroupJoinState is an enum already 🤔

File: src/requests/GroupsApi.ts
Comment (lolmaxz) [reply src/requests/GroupsApi.ts:334]: not exactly because this isn't a code breaking error, it's just bad data, we can handle it if it's out of range, when it's not valid, it will just use the default value. You think it's still better to throw an error despite this?

File: src/requests/UsersApi.ts
Comment (lolmaxz) [reply src/requests/UsersApi.ts:156]: that was an old code from my last comit lol it just got moved along I will redo this lol

File: src/requests/AuthApi.ts
Comment (lolmaxz) [reply src/requests/AuthApi.ts:77]: Fixed

File: src/requests/AuthApi.ts
Comment (lolmaxz) [reply src/requests/AuthApi.ts:58]: tweaked it to be better.

File: src/requests/GroupsApi.ts
Comment (Huijiro) [reply src/requests/GroupsApi.ts:62]: Your GroupJoinState enum is: 
```typescript
export enum GroupJoinState {
    Open = 'open',
    Closed = 'closed',
    Invite = 'invite',
    Request = 'request',
}
```

But here with the way you're using it, it should be:
```typescript
export enum GroupJoinState {
    Closed = 0,
    Invite = 1,
    Request = 2,
    Open = 3,
}
```

Comment (Huijiro) [reply src/requests/GroupsApi.ts:62]: That's why i told you, you should review what [Enums do in Typescript](https://www.typescriptlang.org/docs/handbook/enums.html)

File: src/requests/UsersApi.ts
Comment (lolmaxz) [reply src/requests/UsersApi.ts:156]: > ???? Really?

thinking back on it, there is a very good reason I did it the way I did, I could explain to you by VC, but technically, vrchat users have multiple tags, I need to find the first one the user has. I could maybe optimize this little code but it does what I want it to do in the right order.

File: src/requests/UsersApi.ts
Comment (lolmaxz) [reply src/requests/UsersApi.ts:172]: There is tags from vrchat like "system_trust_known", there is the enum Known and there is the definition "Known User". That's why I did it the way I did.

File: src/requests/UsersApi.ts
Comment (lolmaxz) [reply src/requests/UsersApi.ts:191]: This is completely irelevant, indeed that is how vrchat tags works, there is indeed a tag called nuisance and a tag called troll, for both cases I simply made a function to know if a user has any of the tags, but normally, if a user has one of them, they have the other one, I just played it safe.

File: src/VRCWrapper.ts
Comment (lolmaxz) [reply src/VRCWrapper.ts:198]: I will work soon again on errors handling and redoing my error class a little.

File: src/requests/UsersApi.ts
Comment (lolmaxz) [reply src/requests/UsersApi.ts]: I ended up fixing completely this function as well as the two bellow it 😉

File: src/VRCWrapper.ts
Comment (lolmaxz) [reply src/VRCWrapper.ts:64]: The main VRCWrapper will be reworked a bit maybe tomorrow, I already tweaked it a bit tonight, it's getting more and more clean now.

File: src/requests/GroupsApi.ts
Comment (lolmaxz) [reply src/requests/GroupsApi.ts:1]: all rejects has been removed.

**Comment by lolmaxz** [comment]
I just pushed a bit more changes

**Comment by lolmaxz** [comment]
I believe this commit is good to be pushed for now. good progress and great implementations.

**Pull request merged**
